lib/transproc/recursion.rb

```diff
@@ -30,13 +32,10 @@ module Recursion
     # @api public
     def array_recursion(value, fn)
       result = fn[value]
+      guarded = Transproc(:guard, -> v { v.is_a?(::Array) }, -> v { Transproc(:array_recursion, fn)[v] })
```

Comment:
We could maybe introduce type-check functions. `Transproc.register(:is_array, -> v { v.is_a?(Array) })`?? /cc @splattael